### PR TITLE
#25112 No longer use GetTable to implement CREATE_IF_NEEDED to avoid low quotas

### PR DESCRIPTION
**Pull request by reuvenlax**
https://github.com/apache/beam/issues/25112

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @prodriguezdefino

**Comment by reuvenlax** [comment]
Run Spotless PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTableHelpers.java [hunk 1]
```diff
@@ -46,8 +48,28 @@ public class CreateTableHelpers {
...
+  // When CREATE_IF_NEEDED is specified, BQ tables should be created if they do not exist. This
+  // method detects
+  // errors on table operations, and attempts to create the table if necessary.
+  static void createTableWrapper(Callable<Void> action, Callable<Boolean> tryCreateTable)
+      throws Exception {
+    for (int retry = 0; retry <= 1; ++retry) {
+      try {
+        action.call();
+      } catch (ApiException | StatusRuntimeException e) {
+        // TODO: Once BigQuery reliably returns a consistent error on table not found, we should
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTableHelpers.java:60]: Maybe an issues can be created in the client to track the feature request / improvement.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTableHelpers.java [hunk 2]
```diff
@@ -46,8 +48,28 @@ public class CreateTableHelpers {
...
+        // only try creating
+        // the table on that error.
+        boolean created = tryCreateTable.call();
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTableHelpers.java:63]: not sure about this but, should this backoff before retrying? more considering that we are requesting the table info from BQ before sending the create request.

I think we would want to resolve the table creation as fast as possible but as many bundle instances will try to do the same, almost at the same time, maybe backoff could save some create request if we wait some amount of time on some bundles (not sure, maybe randomly?)

**Review by prodriguezdefino — COMMENTED** [review]
It looks good to me in general, I like the idea of only requesting creating when needed (the table does not exists) instead of checking if we need to do it for every bundle instance creation (avoiding lots of get info requests that may not be necessary). 

The only question I have is that in case of error while trying to create clients/appending/etc. shouldn't we backoff a little bit in some threads? If not maybe we can cause a similar quota exhaustion by having multiple bundles trying to do the same almost at the same time (for large pipelines that could be several dozens per worker).

**Comment by reuvenlax** [comment]
One thing to note: CreateTableHelpers keeps a process-wide static cache of tables known to exist, so it was never done on every bundle. The problem was. that when using at-least-once we had to check every table at least once on every worker, and the quota was too low for that.

**Comment by prodriguezdefino** [comment]
> One thing to note: CreateTableHelpers keeps a process-wide static cache of tables known to exist, so it was never done on every bundle. The problem was. that when using at-least-once we had to check every table at least once on every worker, and the quota was too low for that.

Maybe I misunderstood, but weren't we clearing the destination's map on every bundle start exec? 

```
@StartBundle
     public void startBundle() {
       destinations = Maps.newHashMap();
     }
```

**Comment by reuvenlax** [comment]
The static map is in CreateTableHelpers.java

**Review by prodriguezdefino — APPROVED** [review]
ok, then this applies for this implementation as well, only 1 bundle per worker will actually create the table, the rest will get the cached reference.

**Comment by prodriguezdefino** [comment]
LGTM

**Comment by johnjcasey** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by johnjcasey** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
FYI - testing in production identified a bug in BigQuery that prevent merging of this PR. The BigQuery team is working on fixing this bug, and once this bug is fixed this PR can be merged.

**Comment by johnjcasey** [comment]
@reuvenlax does BQ have an estimate for their fix?

**Comment by johnjcasey** [comment]
It looks like the delay here is prohibitive to include in the release, and it has been an issue for several versions already, so this will be pushed to 2.46

**Comment by damccorm** [comment]
@reuvenlax FYI, the 2.46 cut is a week from today - do you think this will make the cut? If not, I don't think it seems like a release blocker (for the same reasons mentioned above) - does that seem right?

**Comment by yirutang** [comment]
> FYI - testing in production identified a bug in BigQuery that prevent merging of this PR. The BigQuery team is working on fixing this bug, and once this bug is fixed this PR can be merged.

What is the bug?

**Comment by reuvenlax** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
R: @yirutang BigQuery is still failing create stream after the table has been created - presumably still hitting caches. I added some extra retry/backoff here, which seems sufficient to make the test pass.

**Comment by reuvenlax** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
run java precommit

**Pull request merged**